File: bin/make_silico.py
#!/usr/bin/env python
import sys
import subprocess
import re
import os
from Bio import SeqIO
import argparse
from rebase import rebase

# Usage: make_silico.py fasta enzyme outfile

# find restriction sites 
# dna is a contig
def findSites(dna, recSeq):
    sites = []
    lr = len(recSeq)
    strs = (dna[loc:loc+lr] for loc in xrange(len(dna)))
    sites = [i for i,s in enumerate(strs) if s==recSeq]
    return sites
   
def createInsilico(fout, dna, contigId, recSeq):
    ldna = len(dna)
    sites = findSites(dna, recSeq)
    numSites = len(sites)
    if (ldna > 0 and numSites > 0):
        outString = ' '.join([contigId,str(ldna), str(numSites)])+'\n'
        outString += ';'.join(str(site) for site in sites)
        fout.write(outString + '\n')

# Return arguments:
#   fasta, enzyme, outfile
def parseArgs():
    parser = argparse.ArgumentParser(description="Create in-silico optical map from FASTA")
    parser.add_argument('fasta', help = 'Path to input FASTA file')
    parser.add_argument('enzyme', help = 'Enzyme name')
    parser.add_argument('outfile', help = 'Output file.')
    args = parser.parse_args()
    return args

def getRecSeq(enzyme):
   
    enzyme = enzyme.upper()
    # Get the enzyme recognition pattern
    if enzyme not in rebase:
        raise RuntimeError('No Match for enzyme %s in rebase'%enzyme)

    recSeqs = rebase[enzyme]

    if len(recSeqs)>1:
        raise RuntimeError('Found %i matches for enzyme %s\n'%(len(mLines), enzyme))

    recSeq = recSeqs[0]

    # Check that the resctriction_seq is not degenerate
    if(re.search('[^acgtACGT]', recSeq)):
        raise RuntimeError('ERROR: %s has a degenerate recognition sequence %s'%(enzyme, recSeq))

    # Convert recognition sequence to upper case
    recSeq = recSeq.upper()
    return recSeq
   
def makeInsilicoMain(fasta, enzyme, outFile):
    recSeq = getRecSeq(enzyme)
    fin = open(fasta)
    fout = open(outFile, 'w')
    for rec in SeqIO.parse(fin, 'fasta'):
        dna = str(rec.seq).upper()
        contigId = rec.id
        createInsilico(fout, dna, contigId, recSeq)
    fin.close()
    fout.close()

def main():
    args = parseArgs()
    fasta = args.fasta
    enzyme = args.enzyme
    outFile = args.outfile
    makeInsilicoMain(fasta, enzyme, outFile)


if __name__ == '__main__':
    main()

File: bin/make_silico.pl
#!/usr/bin/perl

#
# The second argument can also be an enzyme name that
# can be found in repbase.staden
#
($infile, $restriction_seq, $outfile) = @ARGV;

if($match = `grep $restriction_seq repbase.staden`) {

    ($name, $restriction_seq) = split(/\//, $match);
    $restriction_seq =~ s/\'//g;
}

print "ERROR: make_silico.pl was given the argument $restriction_seq and couldn't find the enzyme OR\n". 
      "       was given a degenerate restriction sequence\n" if($restriction_seq =~ m/[^acgtACGT]/);

$restriction_seq = uc($restriction_seq);

open(IN, $infile);
$lines = join("", <IN>);
close(IN);

open(OUT, ">$outfile");
while($lines =~ m/\>(\S+).*\n([^\>]+)/g) {

    $name = $1; $seq = $2; $seq =~ s/\n//g; $seq = uc($seq);
    
    @positions = (); $pos = 0;
    while(($pos = index($seq, $restriction_seq, $pos+1)) != -1) {

	push @positions, ($pos+1);
    }

    print OUT ("$name ".length($seq)." ".@positions."\n@positions \n") if(@positions != 0);
}

close(OUT);
